GVFS/GVFS.Common/LibGit2RepoPool.cs

```diff
@@ -1,47 +1,66 @@
 ﻿using GVFS.Common.Git;
 using GVFS.Common.Tracing;
 using System;
-using System.Collections.Concurrent;
 using System.Threading;
 
 namespace GVFS.Common
 {
     public class LibGit2RepoPool
     {
-        private const int TryAddTimeoutMilliseconds = 10;
-        private const int TryTakeTimeoutMilliseconds = Timeout.Infinite;
+        private static readonly TimeSpan DefaultRepositoryDisposalPeriod = TimeSpan.FromMinutes(15);
 
-        private readonly BlockingCollection<LibGit2Repo> pool;
+        private readonly Func<LibGit2Repo> createRepo;
         private readonly ITracer tracer;
+        private readonly object sharedRepoLock = new object();
+        private readonly TimeSpan sharedRepositoryDisposalPeriod;
+        private bool stopped;
+        private LibGit2Repo sharedRepo;
+        private Timer sharedRepoDisposalTimer;
+        private int activeCallers;
 
-        public LibGit2RepoPool(ITracer tracer, Func<LibGit2Repo> createRepo, int size)
+        public LibGit2RepoPool(ITracer tracer, Func<LibGit2Repo> createRepo, int size = 1, TimeSpan? disposalPeriod = null)
         {
             if (size <= 0)
             {
                 throw new ArgumentException("ProcessPool: size must be greater than 0");
             }
 
             this.tracer = tracer;
-            this.pool = new BlockingCollection<LibGit2Repo>();
-            for (int i = 0; i < size; ++i)
-            {
-                this.pool.Add(createRepo());
-            }
+            this.createRepo = createRepo;
+
+            this.sharedRepositoryDisposalPeriod = disposalPeriod ?? DefaultRepositoryDisposalPeriod;
+
+            this.sharedRepoDisposalTimer = new Timer(
+                (state) => this.DisposeSharedRepo(),
+                state: null,
+                dueTime: this.sharedRepositoryDisposalPeriod,
+                period: this.sharedRepositoryDisposalPeriod);
         }
 
         public void Dispose()

```

Comment:
I see we have a `Dispose` method already (not `IDisposable.Dispose` though).
Should we not just implement `IDisposable` for good measure? Are the cases when it's legitimate to create an `LibGit2RepoPool` without calling `Dispose`?